packages/runtime/runtime-utils/src/summarizerNodeWithGc.ts

```diff
@@ -0,0 +1,164 @@
+/*!
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { ITelemetryLogger } from "@fluidframework/common-definitions";
+import { assert } from "@fluidframework/common-utils";
+import {
+    IContextSummarizeResult,
+    IGraphNode,
+    ISummarizeInternalResult,
+    ISummarizerNodeConfig,
+    ISummarizerNodeWithGC,
+    CreateChildSummarizerNodeParam,
+} from "@fluidframework/runtime-definitions";
+import { ICreateChildDetails, IInitialSummary, SummarizerNode, SummaryNode } from "./summarizerNode";
+import { cloneGCNodes } from "./garbageCollectionUtils";
+
+/**
+ * Extends the functionality of SummarizerNode to manage this node's garbage collection data:
+ * - It caches the list of GC nodes returned by the summarizeInternal method.
+ * - Gets the initial value of the GC nodes if required.
+ * - Adds the cached list of GC nodes to the result of SummarizerNode's summarize.
+ * - Adds trackState param to summarize. If trackState is false, it bypasses the SummarizerNode and calls
+ *   directly into summarizeInternal method.
+ */
+export class SummarizerNodeWithGC extends SummarizerNode implements ISummarizerNodeWithGC {
+    private gcNodes: IGraphNode[] | undefined;
+    protected constructor(
+        logger: ITelemetryLogger,
+        private readonly summarizeFn: (fullTree: boolean, trackState: boolean) => Promise<ISummarizeInternalResult>,
+        config: ISummarizerNodeConfig,
+        changeSequenceNumber: number,
+        /** Undefined means created without summary */
+        latestSummary?: SummaryNode,
+        initialSummary?: IInitialSummary,
+        wipSummaryLogger?: ITelemetryLogger,
+        /** Function to get the initial value of garbage collection nodes */
+        private readonly getInitialGCNodesFn?: () => Promise<IGraphNode[]>,
+    ) {
+        super(
+            logger,
+            async (fullTree: boolean) => this.summarizeInternal(fullTree, true /* trackState */),
+            config,
+            changeSequenceNumber,
+            latestSummary,
+            initialSummary,
+            wipSummaryLogger,
+        );
+    }
+
+    public async summarize(fullTree: boolean, trackState: boolean = true): Promise<IContextSummarizeResult> {
```

Comment:
Curious: why move `trackState` concept into SummarizerNode?

Reply:
Nevermind, I get it. It's to always do the GC node work even if `trackState` is false. I misread it the first time.